astropy/units/tests/test_units.py

```diff
@@ -87,11 +88,13 @@ def test_convert():
 def test_convert_roundtrip():
     c1 = u.cm.get_converter(u.m)
     c2 = u.m.get_converter(u.cm)
-    np.isclose(c1(c2(10.0)), c2(c1(10.0)), atol=0, rtol=1e-15)
+    assert pytest.approx(c1(c2(10.0)), rel=1e-15, abs=0) == c2(c1(10.0))
```

Comment:
Throughout the astropy test, we tend to just use `np.testing` tools. Let's just use `assert_allclose` here, which is already imported. Fine to set `rtol` and `atol` as you have them here.

Reply:
```python
def test_convert_roundtrip():
    c1 = u.cm.get_converter(u.m)
    c2 = u.m.get_converter(u.cm)
    assert assert_allclose(c1(c2(10.0)), c2(c1(20.0)), rtol=1e-15, atol=0)
```
produces
```
============================================== FAILURES ===============================================
_______________________________________ test_convert_roundtrip ________________________________________

    def test_convert_roundtrip():
        c1 = u.cm.get_converter(u.m)
        c2 = u.m.get_converter(u.cm)
>       assert assert_allclose(c1(c2(10.0)), c2(c1(20.0)), rtol=1e-15, atol=0)

astropy/units/tests/test_units.py:91: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (<function assert_allclose.<locals>.compare at 0x7842243ded40>, array(10.), array(20.))
kwds = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=1e-15, atol=0', 'strict': False, ...}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-15, atol=0
E           
E           Mismatched elements: 1 / 1 (100%)
E           Max absolute difference among violations: 10.
E           Max relative difference among violations: 0.5
E            ACTUAL: array(10.)
E            DESIRED: array(20.)

/usr/lib/python3.12/contextlib.py:81: AssertionError
```
whereas
```python
def test_convert_roundtrip():
    c1 = u.cm.get_converter(u.m)
    c2 = u.m.get_converter(u.cm)
    assert pytest.approx(c1(c2(10.0)), rel=1e-15, abs=0) == c2(c1(20.0))
```
produces
```
============================================== FAILURES ===============================================
_______________________________________ test_convert_roundtrip ________________________________________

    def test_convert_roundtrip():
        c1 = u.cm.get_converter(u.m)
        c2 = u.m.get_converter(u.cm)
>       assert pytest.approx(c1(c2(10.0)), rel=1e-15, abs=0) == c2(c1(20.0))
E       assert 10.0 ± 1.0e-14 == 20.0
E         
E         comparison failed
E         Obtained: 20.0
E         Expected: 10.0 ± 1.0e-14

astropy/units/tests/test_units.py:91: AssertionError
```
There is much less clutter when using `pytest.approx()`, so I don't see why we would want to use `assert_allclose()` with scalar values.

That being said, I should move the `pytest.approx()` calls to the right hand side of the comparisons.

Reply:
> This is not about what looks better, ...

What the output looks like in case of a test failure is important. If it weren't then there would be no reason use `numpy.testing` either, we could just use `numpy.allclose()` instead.

> By introducing `pytest.approx` here you are forcing me and other developers to think how `pytest.approx` even works - and, like much of pytest, it does not work in an obvious way since only one entry is included in the call, ...

Relative tolerances inherently create an asymmetry. If `approx()` helps make that more explicit then that's good.

> ... so I either have to guess it affects how `__eq__` is done or look that up. 

[`approx()`](https://docs.pytest.org/en/8.3.x/reference/reference.html#pytest-approx) has an excellent docstring, and looking up docstrings is simple with any editor that is worth using.

> There's no need to force new developers to learn this, ...

Developers (and not just new ones) would surely benefit from learning how to use `pytest`.

> ...when we already have a standard that we follow the meaning of which is totally clear.

Overall `astropy` tests are very inconsistent and don't really follow any standards.